File: frontend/src/features/room/roomHeightUtils.js
const RANGE_PATTERN = /^([\d.]+)\s*(?:-|–|—|\.\.|to)\s*([\d.]+)$/i;

const toMillimetres = (value) => {
    const number = parseFloat(value);
    if (Number.isNaN(number) || number <= 0) {
        return null;
    }
    // Values <= 50 are treated as metres (e.g. 5 - 10 → 5000-10000 mm).
    return number <= 50 ? number * 1000 : number;
};

export const parseRoomHeightInput = (raw) => {
    if (raw === null || raw === undefined || String(raw).trim() === '') {
        return { ok: false, error: 'Room height is required' };
    }

    if (typeof raw === 'number') {
        if (raw <= 0) {
            return { ok: false, error: 'Room height must be greater than 0' };
        }
        return {
            ok: true,
            height: raw,
            height_min: raw,
            height_max: raw,
        };
    }

    let text = String(raw).trim().toLowerCase();
    text = text.replace(/\bmm\b/g, '').replace(/\bm\b/g, '').trim();

    const rangeMatch = text.match(RANGE_PATTERN);
    if (rangeMatch) {
        const first = toMillimetres(rangeMatch[1]);
        const second = toMillimetres(rangeMatch[2]);
        if (first === null || second === null) {
            return { ok: false, error: 'Enter a valid height range, e.g. 5000-6000 or 5-10 (m)' };
        }
        const height_min = Math.min(first, second);
        const height_max = Math.max(first, second);
        return { ok: true, height: height_max, height_min, height_max };
    }

    const single = toMillimetres(text);
    if (single === null) {
        return { ok: false, error: 'Enter a valid height, e.g. 3000 or 5000-6000' };
    }

    return {
        ok: true,
        height: single,
        height_min: single,
        height_max: single,
    };
};

export const formatRoomHeightForInput = (room) => {
    if (!room) return '';
    const min = room.height_min ?? room.height;
    const max = room.height_max ?? room.height;
    if (min != null && max != null && min !== max) {
        return `${min} - ${max}`;
    }
    if (room.height != null && room.height !== '') {
        return String(room.height);
    }
    return '';
};

export const formatRoomHeightLabel = (room) => {
    if (!room) return 'EXT. HT. No height';
    const min = room.height_min ?? room.height;
    const max = room.height_max ?? room.height;
    if (min != null && max != null && min !== max) {
        return `EXT. HT. ${min}-${max}mm`;
    }
    if (room.height != null && room.height !== '') {
        return `EXT. HT. ${room.height}mm`;
    }
    return 'EXT. HT. No height';
};

export const isRoomHeightRange = (room) => {
    const min = room?.height_min ?? room?.height;
    const max = room?.height_max ?? room?.height;
    return min != null && max != null && min !== max;
};

File: core/room_height_utils.py
import re


def _to_mm(value):
    """Convert a numeric input to mm; values <= 50 are treated as metres."""
    number = float(value)
    if number <= 0:
        raise ValueError('Height must be greater than 0')
    if number <= 50:
        return number * 1000.0
    return number


def parse_height_input(raw):
    """
    Parse a room height string.
    Accepts: '3000', '5000-6000', '5 - 10' (metres), '5m - 10m'.
    Returns dict with height, height_min, height_max (all in mm).
    """
    if raw is None or (isinstance(raw, str) and not raw.strip()):
        raise ValueError('Room height is required')

    if isinstance(raw, (int, float)):
        height = float(raw)
        if height <= 0:
            raise ValueError('Room height must be greater than 0')
        return {'height': height, 'height_min': height, 'height_max': height}

    text = str(raw).strip().lower()
    text = re.sub(r'\bmm\b', '', text)
    text = re.sub(r'\bm\b', '', text)
    text = text.strip()

    range_match = re.match(
        r'^([\d.]+)\s*(?:-|–|—|\.\.|to)\s*([\d.]+)$',
        text,
        flags=re.IGNORECASE,
    )
    if range_match:
        minimum = _to_mm(range_match.group(1))
        maximum = _to_mm(range_match.group(2))
        if minimum > maximum:
            minimum, maximum = maximum, minimum
        return {'height': maximum, 'height_min': minimum, 'height_max': maximum}

    height = _to_mm(text)
    return {'height': height, 'height_min': height, 'height_max': height}


def normalize_room_height_fields(room_data):
    """Ensure height, height_min, and height_max are consistent on room payloads."""
    if room_data.get('height_min') is not None and room_data.get('height_max') is not None:
        minimum = float(room_data['height_min'])
        maximum = float(room_data['height_max'])
        if minimum > maximum:
            minimum, maximum = maximum, minimum
        room_data['height_min'] = minimum
        room_data['height_max'] = maximum
        room_data['height'] = maximum
        return room_data

    if room_data.get('height') is not None:
        if isinstance(room_data['height'], str):
            parsed = parse_height_input(room_data['height'])
            room_data.update(parsed)
            return room_data
        height = float(room_data['height'])
        room_data['height'] = height
        room_data['height_min'] = height
        room_data['height_max'] = height

    return room_data
